src/main/java/ch/njol/skript/registrations/EventValues.java

```diff
@@ -19,224 +19,351 @@
+		private final Class<E> event;
+		private final Class<T> type;
+
+		private int priority;
+
+		@SuppressWarnings("unchecked")
+		public EventValueInfo(Class<E> event, Class<T> type, Getter<T, E> getter, int priority, @Nullable String excludeErrorMessage, @Nullable Class<? extends E>... excludes) {
 			assert getter != null;
-			this.event = event;
-			this.c = c;
-			this.getter = getter;
-			this.excludes = excludes;
+			assert event != null;
+			assert type != null;
+
+			this.excludes = excludes == null ? new Class[0] : excludes;
+			if (excludes.length > 0 && excludeErrorMessage == null)
+				Skript.error("The exclude error message cannot be null when there are excluded events.");
+
 			this.excludeErrorMessage = excludeErrorMessage;
+			this.priority = priority;
+			this.getter = getter;
+			this.event = event;
+			this.type = type;
 		}
-		
-		/**
-		 * Get the class that represents the Event.
-		 * @return The class of the Event associated with this event value
-		 */
-		public Class<E> getEventClass() {
-			return event;
-		}
-		
-		/**
-		 * Get the class that represents Value.
-		 * @return The class of the Value associated with this event value
-		 */
-		public Class<T> getValueClass() {
-			return c;
+
+		@Override
+		public int compareTo(EventValueInfo<?, ?> other) {
+			return Integer.compare(Math.abs(priority), Math.abs(other.priority));
 		}
-		
-		/**
-		 * Get the classes that represent the excluded for this Event value.
-		 * @return The classes of the Excludes associated with this event value
-		 */
-		@Nullable
-		@SuppressWarnings("null")
-		public Class<? extends E>[] getExcludes() {
-			if (excludes != null)
-				return Arrays.copyOf(excludes, excludes.length);
-			return new Class[0];
+
+		@Override
+		public String toString() {
+			return "(" + event.getSimpleName() + "#" + type.getSimpleName() + " at priority " + priority + ")";
 		}
-		
-		/**
-		 * Get the error message used when encountering an exclude value.
-		 * @return The error message to use when encountering an exclude
-		 */
-		@Nullable
-		public String getExcludeErrorMessage() {
-			return excludeErrorMessage;
+
+		@Override
+		public boolean equals(Object object) {
+			if (!(object instanceof EventValueInfo))
+				return false;
+			EventValueInfo<?, ?> other = (EventValueInfo<?, ?>) object;
+			if (!event.equals(other.event))
+				return false;
+			if (!type.equals(other.type))
+				return false;
+			if (!getter.equals(other.getter))
+				return false;
+			if (priority != other.priority)
+				return false;
+			if (!excludes.equals(other.excludes))
+				return false;
+			return true;
 		}
+
 	}
-	
-	private final static List<EventValueInfo<?, ?>> defaultEventValues = new ArrayList<>(30);
+
+	private final static List<EventValueInfo<?, ?>> defaultEventValues = new ArrayList<>();
 	private final static List<EventValueInfo<?, ?>> futureEventValues = new ArrayList<>();
 	private final static List<EventValueInfo<?, ?>> pastEventValues = new ArrayList<>();
```

Comment:
If these need sorted and can't have duplicates perhaps it would be better to use a TreeSet?

Reply:
I already tried that, it wasn't working and I tried everything. The sets ended up having only 4-6 values for some reason.